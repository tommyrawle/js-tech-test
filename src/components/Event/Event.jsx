import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import * as actions from '../../redux/actions';
import { formatDateOrTime } from '../../redux/utils';
import { getEventDetails, getMarketDetails } from '../../redux/getters';
import { getDisplayableEvents, getSortedMarketIds } from '../../redux/selectors';
import Spinner from '../Spinner/Spinner.jsx';

import {
  EventMarkets,
  EventDetailsBlock,
  EventDetailsTeam,
  EventDetailsPosition,
  EventDetailsTeamName,
  EventDetailsScore,
  SecondaryDetailsBlock,
  DetailsEventType,
  DetailsEventDate,
  LiveIndicator
} from './Event.styles.jsx';
import Market from '../Market/Market.jsx';

const Event = ({ events, loading, setEvent, setMarkets, sortedMarketIds, setLoadingStatus }) => {
  const eventId = parseInt(useParams().eventId);

  const getEventPageData = async eventId => {
    setLoadingStatus(true);
    const event = await getEventDetails(eventId);
    setEvent(event);
    Promise.all(event.markets.map(async market => await getMarketDetails(market))).then(marketData => {
      setMarkets(marketData);
    });
  };

  useEffect(() => {
    getEventPageData(eventId);
  }, []);

  if (loading) return <Spinner />;
  else {
    const event = events.find(event => event.eventId === eventId);
    let teams = new Object();
    event.competitors.forEach(team => {
      teams[team.position] = team.name;
    });

    return (
      <div>
        <EventDetailsBlock>
          <EventDetailsTeam alignment={'right'}>
            <EventDetailsPosition>Home</EventDetailsPosition>
            <EventDetailsTeamName>{teams.home}</EventDetailsTeamName>
            <EventDetailsScore>{event.scores.home}</EventDetailsScore>
          </EventDetailsTeam>
          <EventDetailsTeam alignment={'left'}>
            <EventDetailsPosition>Away</EventDetailsPosition>
            <EventDetailsTeamName>{teams.away}</EventDetailsTeamName>
            <EventDetailsScore>{event.scores.away}</EventDetailsScore>
          </EventDetailsTeam>
        </EventDetailsBlock>

        <SecondaryDetailsBlock>
          <DetailsEventType>
            {event.linkedEventTypeName ? `${event.linkedEventTypeName} / ${event.typeName}` : event.typeName}
          </DetailsEventType>
          <DetailsEventDate>
            {formatDateOrTime(event.startTime, 'date')} | {formatDateOrTime(event.startTime, 'time')}
            {event.status.live && <LiveIndicator />}
          </DetailsEventDate>
        </SecondaryDetailsBlock>

        <EventMarkets>
          {sortedMarketIds.map((marketId, index) => {
            return <Market key={index} marketId={marketId} outcomeOpen={index < 10} />;
          })}
        </EventMarkets>
      </div>
    );
  }
};

Event.propTypes = {
  events: PropTypes.arrayOf(PropTypes.object),
  markets: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool
};

const mapStateToProps = state => ({
  events: getDisplayableEvents(state),
  sortedMarketIds: getSortedMarketIds(state),
  loading: state.loading
});

const mapDispatchToProps = dispatch => ({
  setEvent: eventObject => dispatch(actions.setEvent(eventObject)),
  setMarkets: arrayOfMarkets => dispatch(actions.setMarkets(arrayOfMarkets)),
  setLoadingStatus: bool => dispatch(actions.setLoadingStatus(bool))
});

export default connect(mapStateToProps, mapDispatchToProps)(Event);
