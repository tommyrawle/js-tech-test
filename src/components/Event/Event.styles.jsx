import styled from 'styled-components';

export const EventListTableCell = styled.td`
  padding: ${({ eventCell }) => (eventCell ? '10px 20px' : '')};
  border-bottom: 1px solid #ececec;
`;

export const EventMarkets = styled.div`
  width: 60%;
  margin: 0 auto;
  @media (max-width: 676px) {
    width: 100%;
  }
`;
