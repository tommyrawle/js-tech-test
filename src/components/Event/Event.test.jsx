import React from 'react';
import { shallow } from 'enzyme';
import { Event } from './Event.jsx';
import '@babel/polyfill';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    eventId: '21249939'
  })
}));

describe('Event component', () => {
  it('should render Event component', () => {
    const mockProps = {
      events: [
        {
          eventId: 21249939,
          name: 'Shanghai Shenhua 0 v 0 Shandong Luneng Taishan',
          typeName: 'Football Live',
          linkedEventTypeName: 'Chinese Super League',
          startTime: '2017-09-19T11:35:23.000Z',
          scores: { home: 0, away: 0 },
          competitors: [
            { name: 'Shanghai Shenhua', position: 'home' },
            { name: 'Shandong Luneng Taishan', position: 'away' }
          ],
          status: {
            live: true,
            displayable: true
          }
        }
      ],
      loading: false,
      setEvent: jest.fn(),
      setMarkets: jest.fn(),
      sortedMarketIds: [123],
      setLoadingStatus: jest.fn(),
      eventId: 21249939
    };
    const wrapper = shallow(<Event {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
