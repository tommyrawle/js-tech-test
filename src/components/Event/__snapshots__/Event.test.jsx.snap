// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event component should render Event component 1`] = `
<div>
  <styled.div>
    <styled.div
      alignment="right"
    >
      <styled.div>
        Home
      </styled.div>
      <styled.div>
        Shanghai Shenhua
      </styled.div>
      <styled.div>
        0
      </styled.div>
    </styled.div>
    <styled.div
      alignment="left"
    >
      <styled.div>
        Away
      </styled.div>
      <styled.div>
        Shandong Luneng Taishan
      </styled.div>
      <styled.div>
        0
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.h2>
      Chinese Super League / Football Live
    </styled.h2>
    <styled.div>
      Tuesday 19 September 2017
       | 
      12:35
      <styled.span />
    </styled.div>
  </styled.div>
  <styled.div>
    <Connect(Market)
      key="0"
      marketId={123}
      outcomeOpen={true}
    />
  </styled.div>
</div>
`;
